File: 2022/haskell/day10.hs
-- Advent of Code 2022, day 10
--  Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  part2 xs

-- Parsing the input

data Inst = Noop | Addx Int | NextAddx Int
  deriving (Show,Eq)

pData :: Parser Inst
pData = do symbol "noop"
           return Noop
        <|>
        do symbol "addx"
           x <- integer
           return (Addx x)

pInput :: Parser [Inst]
pInput = many pData

-- Part 1

cycleTest :: Int -> Bool
cycleTest cycle = (cycle - 20) `mod` 40 == 0

strUp :: Int -> Int -> Int -> Int
strUp cycle reg str =
  if cycleTest cycle then str + cycle*reg else str

exec :: [Inst] -> (Int,Int,Int) -> Int
exec [] (cycle,reg,str) = str
exec (Noop:insts) (cycle,reg,str) =
   exec insts (cycle+1, reg, strUp cycle reg str)
exec (Addx x:insts) (cycle,reg,str) =
   exec (NextAddx x:insts) (cycle+1, reg, strUp cycle reg str)
exec (NextAddx x:insts) (cycle,reg,str) =
   exec insts (cycle+1, reg+x, strUp cycle reg str)

part1 :: [Inst] -> Int
part1 insts = exec insts (1,1,0)

-- Part 2

pixel :: Int -> Int -> Char
pixel cycle reg =
  if abs ((cycle-1) `mod` 40 - reg) <= 1 then '#' else '.'

execCRT :: [Inst] -> (Int,Int) -> String
execCRT [] (cycle,reg) = ""
execCRT (Noop:insts) (cycle,reg) =
   pixel cycle reg : execCRT insts (cycle+1, reg)
execCRT (Addx x:insts) (cycle,reg) =
   pixel cycle reg : execCRT (NextAddx x:insts) (cycle+1, reg)
execCRT (NextAddx x:insts) (cycle,reg) =
   pixel cycle reg : execCRT insts (cycle+1, reg+x)

printCRT :: String -> IO ()
printCRT "" = return ()
printCRT crt = do let (ln,crt') = splitAt 40 crt
                  putStrLn ln
                  printCRT crt'
                  
part2 :: [Inst] -> IO ()
part2 insts = printCRT (execCRT insts (1,1))

File: 2022/python/day10.py
# Advent of Code 2022, day 10
#   Venanzio Capretta

print("Advent of Code 2022, Day 10")

f = open("../input10")
input = f.read().splitlines()
f.close()

# Parsing the input

program = []
for s in input:
  if s[0:4] == 'noop':
    program.append(('noop',0))
  else:
    program.append(('addx',int(s[4:])))

# Part 1

actions = [0]
for (inst,x) in program:
  actions.append(0)
  if inst == 'addx':
    actions.append(x)

def signal_strength(c,x):
  return c*x

registerX = 1
sig_sum = 0
for cycle in range(len(actions)):
  if (cycle-20)%40 == 0:
    sig_sum += signal_strength(cycle,registerX)
  registerX += actions[cycle]

print("Part 1: ")
print(sig_sum)

# Part 2

registerX = 1
crt = []
for i in range(6):
  crt_row = []
  for j in range(40):
    cycle = 40*i+j+1
    if abs(j-registerX) <= 1:
      crt_row.append('#')
    else:
      crt_row.append('.')
    registerX += actions[cycle]
  crt.append(crt_row)

print("Part 2: ")
for i in range(6):
  print(''.join(crt[i]))

